kvdb-rocksdb/src/iter.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This module contains an implementation of a RocksDB iterator
+//! wrapped inside a `RwLock`. Since `RwLock` "owns" the inner data,
+//! we're using `owning_ref` to work around the borrowing rules of Rust.
+
+use crate::DBAndColumns;
+use owning_ref::{OwningHandle, StableAddress};
+use parking_lot::RwLockReadGuard;
+use rocksdb::{DBIterator, Direction, IteratorMode};
+use std::ops::{Deref, DerefMut};
+
+pub type KeyValuePair = (Box<[u8]>, Box<[u8]>);
+
+pub struct ReadGuardedIterator<'a, I, T> {
+	inner: OwningHandle<UnsafeStableAddress<RwLockReadGuard<'a, Option<T>>>, DerefWrapper<Option<I>>>,
+}
+
+// We can't implement `StableAddress` for a `RwLockReadGuard`
+// directly due to orphan rules.
+#[repr(transparent)]
+struct UnsafeStableAddress<T>(T);
```

Comment:
```suggestion
struct UnsafeStableAddress<'a, T>(RwLockReadGuard<'a, T>);
```

To make sure no one is doing anything bad ^^